File: latency_test_client.js
// latency_test_client.js

const net = require('net');
const process = require('process');

// --- Configuration ---
const HOST = '127.0.0.1';
const PORT = 65432;
// Use command line argument for NUM_CLIENTS, default to 10 if not provided
const NUM_CLIENTS = parseInt(process.argv[2], 10) || 10;
const DELAY_BETWEEN_CLIENTS = 100; // Time to pause before starting the next client (in milliseconds)
const HOLD_TIME_SECONDS = 5;       // Time to keep all clients connected before disconnection

// --- Global Data Structures ---
const activeSockets = [];
let successfulConnections = 0;

// --- Utility Function: Sleep ---
/**
 * Pauses execution for a specified duration.
 * @param {number} ms - Milliseconds to sleep.
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// --- NEW UTILITY FUNCTION: Clear Socket Buffer ---
/**
 * Clears any buffered, unread data from the socket.
 * This is crucial for latency testing to ensure the response isn't
 * already sitting in the buffer from a prior broadcast.
 * @param {net.Socket} socket
 * @returns {Promise<number>} - The number of messages cleared.
 */
async function clearSocketBuffer(socket) {
    // console.log("[DEBUG] Clearing socket buffer for test client...");
    let clearedCount = 0;
    
    // We'll use a Promise and a temporary 'data' listener that resolves 
    // after a short timeout, ensuring all immediately available data is read.
    return new Promise(async (resolve) => {
        let isClearing = true;
        
        const tempDrainer = (data) => {
            if (isClearing) {
                // console.log(`[CLEARED] ${data.toString().trim()}`);
                clearedCount++;
            }
        };

        socket.on('data', tempDrainer);

        // Give the event loop a chance to process any buffered data
        await sleep(10); // Sleep for 10ms (equivalent to Ruby's 0.01s read_timeout)

        isClearing = false;
        socket.removeListener('data', tempDrainer);
        // console.log(`[DEBUG] Cleared ${clearedCount} buffered messages.`);
        resolve(clearedCount);
    });
}


// --- Core Connection Logic for a Single Client ---
/**
 * Connects a client, sends the registration username, and stores the socket.
 * @param {number} clientId - The sequential ID of the client.
 * @returns {Promise<boolean>} - True if connection was successful, false otherwise.
 */
function connectAndStore(clientId) {
    return new Promise((resolve) => {
        const username = `SeqHold_${clientId}`;
        const client = new net.Socket();

        // 1. Establish Connection
        client.connect(PORT, HOST, () => {
            // 2. Send Registration (Username)
            client.write(username);
            
            // 3. Store the active, registered socket
            activeSockets.push(client);
            successfulConnections++;
            
            // Keep the socket active, but do nothing with incoming data for now.
            client.on('data', (data) => { /* ignore */ }); 
            
            // Handle unexpected disconnection
            client.on('end', () => {
                // Remove socket from the list if it closes unexpectedly
                const index = activeSockets.indexOf(client);
                if (index > -1) {
                    activeSockets.splice(index, 1);
                }
            });

            resolve(true);
        });

        // Handle connection errors
        client.on('error', (err) => {
            if (err.code === 'ECONNREFUSED') {
                console.error(`[Client ${clientId} ERROR] Connection failed (Server offline?).`);
            } else {
                console.error(`[Client ${clientId} ERROR] An error occurred: ${err.message}`);
            }
            client.destroy();
            resolve(false);
        });
    });
}

// --- Main Execution ---
async function startSequentialHoldTest() {
    console.log(`Starting Sequential Hold Test: Connecting ${NUM_CLIENTS} clients one-by-one.`);

    // 1. Sequential Connection Phase
    const connectionStartTime = Date.now();
    for (let i = 1; i <= NUM_CLIENTS; i++) {
        await connectAndStore(i);
        // Pause before starting the next client
        await sleep(DELAY_BETWEEN_CLIENTS);
    }
    const connectionEndTime = Date.now();
    const connectionDuration = (connectionEndTime - connectionStartTime) / 1000;
    
    console.log("\n========================================================");
    console.log("        TEST PHASE 1: CONNECTION SUMMARY         ");
    console.log("========================================================");
    console.log(`Target Clients: ${NUM_CLIENTS}`);
    console.log(`Successful Connections: **${successfulConnections}**`);
    console.log(`Connection Duration: ${connectionDuration.toFixed(2)} seconds`);
    console.log("========================================================");

    if (successfulConnections === 0) {
        console.log("\n[EXIT] No clients connected. Test aborted.");
        return;
    }

    // 2. Hold Phase
    console.log("\n\n========================================================");
    console.log("          TEST PHASE 2: HOLDING CONNECTIONS           ");
    console.log("========================================================");
    console.log(`\n[HOLD] Holding all ${successfulConnections} connections open for **${HOLD_TIME_SECONDS} seconds**...`);
    console.log("Observe the server console for active connection count.");
    await sleep(HOLD_TIME_SECONDS * 1000);

    // 3. Latency Test Phase
    console.log("\n\n========================================================");
    console.log("        TEST PHASE 3: LATENCY OF A COMMAND         ");
    console.log("========================================================");
    
    // Select the first connected socket for the test
    const testSocket = activeSockets[0];
    
    // Clear any messages broadcasted during the HOLD phase
    await clearSocketBuffer(testSocket);
    
    let response = null;
    const timeBeforePrompt = Date.now();
    
    // Send the LOOK command
    testSocket.write("LOOK");
    
    // Set up a listener to capture the next response and measure time
    await new Promise((resolve) => {
        const responseListener = (data) => {
            const message = data.toString().trim();
            if (message.startsWith("Players in")) {
                response = message;
                testSocket.removeListener('data', responseListener);
                resolve();
            }
        };
        
        // Temporarily override the silent listener with one that captures the response
        testSocket.removeAllListeners('data');
        testSocket.on('data', responseListener);
        
        // Set a timeout in case the server fails to respond
        setTimeout(() => {
            if (response === null) {
                testSocket.removeListener('data', responseListener);
                resolve(); // Resolve the promise to continue, but response remains null
            }
        }, 3000); // 3-second timeout for response
    });
    
    const timeAfterPrompt = Date.now();

    if (response === null || !response.startsWith("Players in")) {
        console.error("\n[ERROR] Server disconnected or sent an unexpected response.");
        console.log(`Received: ${response}`);
        var latencySeconds = "N/A (Error)";
    } else {
        var latencySeconds = (timeAfterPrompt - timeBeforePrompt) / 1000;
        console.log(`Latency of Look command: **${latencySeconds.toFixed(9)} seconds**`);
    }

    // 4. Disconnection Phase
    console.log("\n[DISCONNECT] Starting graceful shutdown of all clients...");
    
    // Disconnect all sockets one after the other
    for (let i = 0; i < NUM_CLIENTS; i++) {
        const socket = activeSockets[i];
        try {
            // Send the QUIT command and close the socket.
            socket.write("QUIT");
            socket.end(); // Send FIN packet
            await sleep(DELAY_BETWEEN_CLIENTS);
        } catch (e) {
            // Handle potential errors if the server already closed a socket
        }
    }
    
    console.log(`[COMPLETE] All ${successfulConnections} clients disconnected.`);
    console.log("Test finished.");
} // <-- This brace must be present to close startSequentialHoldTest

// Main execution
if (require.main === module) {
    if (NUM_CLIENTS <= 0) {
        console.error("Please provide a valid number of clients (e.g., node latency_test_client.js 50).");
        process.exit(1);
    }
    // Corrected function call:
    startSequentialHoldTest();
}

File: latency_test_client.rb
require 'socket'
require 'thread'
require 'time'
require 'timeout'

# --- Configuration ---
HOST = '127.0.0.1'
PORT = 65432
NUM_CLIENTS = ARGV[0].to_i         # The requested number of clients to connect sequentially
DELAY_BETWEEN_CLIENTS = 0.1     # Time to pause before starting the next client (in seconds)
HOLD_TIME_SECONDS = 5            # Time to keep all clients connected before disconnection

# --- Global Data Structures ---
$active_sockets = []
$sockets_mutex = Mutex.new

# --- NEW UTILITY FUNCTION: Clear Socket Buffer ---
def clear_socket_buffer(socket)
    puts "[DEBUG] Clearing socket buffer for test client..."
    # Set a very short non-blocking timeout for reading.
    read_timeout = 0.01 
    cleared_count = 0
    
    loop do
        begin
            # Wait for any data to be readable on the socket with a timeout
            readable, _, _ = IO.select([socket], nil, nil, read_timeout)
            
            # If nothing is readable within the timeout, break the loop
            break if readable.nil? || readable.empty?

            # If the socket is readable, read the message
            message = socket.gets
            break unless message # Should only happen if the socket is closed unexpectedly

            # puts "[CLEARED] #{message.chomp}"
            cleared_count += 1
            
        rescue IOError, Errno::ECONNRESET, Timeout::Error
            # Stop clearing if an error occurs (socket closed, etc.)
            break
        end
    end
    puts "[DEBUG] Cleared #{cleared_count} buffered messages."
    return cleared_count
end

# --- Core Connection Logic for a Single Client ---
def connect_and_store(client_id)
    username = "SeqHold_#{client_id}"
    client = nil

    # 1. Establish Connection
    begin
        client = TCPSocket.new(HOST, PORT)
        
        # 2. Send Registration (Username)
        client.puts username

        # 3. Store the active, registered socket (Thread-safe)
        $sockets_mutex.synchronize do
            $active_sockets << client
        end
        
        # We do not attempt to read from the socket to prevent the script from hanging.
        #puts "[Client #{client_id}] Connected and registered as #{username}. Active: #{$active_sockets.size}"
        return true
        
    rescue Errno::ECONNREFUSED
        puts "[Client #{client_id} ERROR] Connection failed (Server offline?)."
    rescue => e
        puts "[Client #{client_id} ERROR] An error occurred: #{e.message}"
    end
    
    # Ensure the socket is closed if connection fails early
    client.close if client
    return false
end

# --- Main Execution ---
def start_sequential_hold_test
    puts "Starting Sequential Hold Test: Connecting #{NUM_CLIENTS} clients one-by-one."
    
    successful_connections = 0
    connection_start_time = Time.now

    # 1. Sequential Connection Phase
    (1..NUM_CLIENTS).each do |i|
        # Connect and store one client
        if connect_and_store(i)
            successful_connections += 1
        end
        
        # Pause before starting the next client
        sleep(DELAY_BETWEEN_CLIENTS)
    end

    connection_end_time = Time.now
    connection_duration = (connection_end_time - connection_start_time).round(2)
    
    puts "\n========================================================"
    puts "        TEST PHASE 1: CONNECTION SUMMARY         "
    puts "========================================================"
    puts "Target Clients: #{NUM_CLIENTS}"
    puts "Successful Connections: **#{successful_connections}**"
    puts "Connection Duration: #{connection_duration} seconds"
    puts "========================================================"

    # 2. Hold Phase
    puts "\n\n========================================================"
    puts "          TEST PHASE 2: HOLDING CONNECTIONS           "
    puts "========================================================"
    puts "\n[HOLD] Holding all #{successful_connections} connections open for **#{HOLD_TIME_SECONDS} seconds**..."
    puts "Observe the server console for active connection count."
    sleep(HOLD_TIME_SECONDS)
 
    test_socket = nil
    $sockets_mutex.synchronize do
        test_socket = $active_sockets[0]
    end
    clear_socket_buffer(test_socket)
    time_before_prompt = Time.now
    test_socket.puts "LOOK"
    response = nil
    begin
        response = test_socket.gets
        time_after_prompt = Time.now
    rescue => e
        puts "[ERROR] Failed to receive response: #{e.message}"
    end
    if response.nil? || !response.start_with?("Players in")
        puts "\n[ERROR] Server disconnected or sent an unexpected response."
        puts "Received: #{response.chomp if response}"
    end
    latency_seconds = time_after_prompt - time_before_prompt

    puts "\n\n========================================================"
    puts "        TEST PHASE 3: LATENCY OF A COMMAND         "
    puts "========================================================"
    puts "Latency of Look command: #{latency_seconds}"
    sleep(0.1)


    # 3. Disconnection Phase
    puts "\n[DISCONNECT] Starting graceful shutdown of all clients..."
    
    # Use a thread-safe copy of the sockets for iteration
    sockets_to_close = []
    $sockets_mutex.synchronize { sockets_to_close.concat($active_sockets) }
    
    # Disconnect all sockets one after the other
    sockets_to_close.each_with_index do |socket, index|
        begin
            # Send the QUIT command and close the socket.
            socket.puts "QUIT" 
            socket.close
            sleep(DELAY_BETWEEN_CLIENTS)
        rescue => e
            # Handle potential errors if the server already closed a socket
        end
    end
    
    puts "[COMPLETE] All #{successful_connections} clients disconnected."
    puts "Test finished."
end

# Main execution
if __FILE__ == $0
    start_sequential_hold_test
end